Add tool for finding Educator Zombie Stories

mjgpy3 opened the pull request:
Goal: give us a signal when a task's taking awhile to get us to discuss the task. E.g. We could decided to pair to or talk through blockers etc... This is all about starting the conversations.

Here's the process changelog that's driving this: https://renaissancelearning.atlassian.net/wiki/spaces/FTE/pages/42818306169/2023-04-24+Push+Model+of+Getting+Help

**Terminology note** Asana calls events that happen to a task "stories". This is very confusing so I'm trying to use "task" throughout to talk about Asana tasks and what our team calls "tasks" or "stories".

This was first built around the cycle-time tool: https://github.com/freckle/asana/tree/main/cycle-time

General algorithm:
 - Parse iterations by name ("Educator N")
 - Take last 5 (may take less if there are gaps for some reason)
 - Get all tasks for those iterations
 - Get all stories (asana events) for those tasks
 - Find ICs by looking at assignees (this might be wrong sometimes... Generally I'm noticing that if it's in the iteration, it's assigned to an IC)
 - Find "done" tasks (those with a start and an end)
 - Find "WIP" tasks (those with just a start)
 - Average how many days "done" tasks take, per point value
 - Find "WIP" tasks that are taking more days than the average for their points

Figuring out the _start_ date is the hard part... The heuristic I've used it to find the first "story" (asana's name for an event) that reflects
 - ~~the task being assigned to a known IC or~~ (no longer consider this due to slack discussion)
 - the task being moved into a known iteration

Along the way if something "fails" (e.g. can't find a start date, assigned to someone else), it's simple excluded from the computations/checks.

Here's the example output at this time (only one zombie ATM)
```shell
$ stack run educator-zombie-tasks
2023-04-25 16:35:16 [info     ] Starting app
2023-04-25 16:35:16 [info     ] Found iterations: [182,183,184,185,186,187,188,189,190,191,192,193,194,195,196,197,198,199,200,201,202,203,204,205,206,207,208,209,210,211,212,213,214,215,216,217,218,219]
2023-04-25 16:35:16 [info     ] Latest iteration                iterationNumber=219
2023-04-25 16:35:16 [info     ] Iterations to search            iterationNumbers=[215, 216, 217, 218, 219]
2023-04-25 16:35:21 [info     ] Relevant tasks                  doneTasksInIterations=39 icsInIterations=[...] tasksWithStories=43 totalTasksInIterations=43 wipTasksInIterations=3
2023-04-25 16:35:21 [info     ] WIP Task                        id=1204285486850983 name=A/B test surface challenging questions on Assignments report  roughStartedAt=2023-04-19T16:04:00.357Z
2023-04-25 16:35:21 [info     ] WIP Task                        id=1203989018545059 name=Optimize `getContentAreaSkillsR` roughStartedAt=2023-04-06T18:29:32.876Z
2023-04-25 16:35:21 [info     ] WIP Task                        id=1203963501833948 name=As a teacher in the treatment group, I can see a recommendation for assigning PreTP to students who need it in the TP flow roughStartedAt=2023-04-19T18:07:11.833Z
2023-04-25 16:35:21 [info     ] Average days to complete points 0.0 point(s)=2 day(s) 1.0 point(s)=2 day(s) 2.0 point(s)=4 day(s) 3.0 point(s)=8 day(s) 5.0 point(s)=13 day(s)
2023-04-25 16:35:21 [info     ] Zombie Task                     averageDaysForPoints=8 daysWip=18 id=1203989018545059 name=Optimize `getContentAreaSkillsR`
```

There should probably be a follow up to get some of these things into our `asana` package (e.g. orphans, story-stuff).

mjgpy3 committed:
Add tool for finding Educator Zombie Stories

Here's the example output at this time
```shell
$ stack run educator-zombie-tasks
2023-04-25 16:35:16 [info     ] Starting app
2023-04-25 16:35:16 [info     ] Found iterations: [182,183,184,185,186,187,188,189,190,191,192,193,194,195,196,197,198,199,200,201,202,203,204,205,206,207,208,209,210,211,212,213,214,215,216,217,218,219]
2023-04-25 16:35:16 [info     ] Latest iteration                iterationNumber=219
2023-04-25 16:35:16 [info     ] Iterations to search            iterationNumbers=[215, 216, 217, 218, 219]
2023-04-25 16:35:21 [info     ] Relevant tasks                  doneTasksInIterations=39 icsInIterations=Thomas Carstens, Cristina Grant, Matthew Pagan, Michael Gilliland, Steven Paul Leiva tasksWithStories=43 totalTasksInIterations=43 wipTasksInIterations=3
2023-04-25 16:35:21 [info     ] WIP Task                        id=1204285486850983 name=A/B test surface challenging questions on Assignments report  roughStartedAt=2023-04-19T16:04:00.357Z
2023-04-25 16:35:21 [info     ] WIP Task                        id=1203989018545059 name=Optimize `getContentAreaSkillsR` roughStartedAt=2023-04-06T18:29:32.876Z
2023-04-25 16:35:21 [info     ] WIP Task                        id=1203963501833948 name=As a teacher in the treatment group, I can see a recommendation for assigning PreTP to students who need it in the TP flow roughStartedAt=2023-04-19T18:07:11.833Z
2023-04-25 16:35:21 [info     ] Average days to complete points 0.0 point(s)=2 day(s) 1.0 point(s)=2 day(s) 2.0 point(s)=4 day(s) 3.0 point(s)=8 day(s) 5.0 point(s)=13 day(s)
2023-04-25 16:35:21 [info     ] Zombie Task                     averageDaysForPoints=8 daysWip=18 id=1203989018545059 name=Optimize `getContentAreaSkillsR`
```

cristygrant commented:
curious, why last 5 iterations? as opposed to last 2 (for carry over calculation)

mjgpy3 commented:
@cristygrant great question! We want to get an average per point value. It's very possible that we, say, only worked on a single 5 pointer in the last couple of iterations so that's not a great sample size!

I'm open to any suggestions if you've got a better idea of a range. 5 was my guess at how many you'd want to actually get some meaningful averages.

cristygrant commented:
oh I see... initially I thought you were trying to find zombie tasks from last 5 iterations and use a "rule of thumb" of 1 pt per day... but I like the approach of getting an average per point value better 👍

mjgpy3 committed:
Change "started" algorithm to just consider "added to iteration"

Also, since the set isn't really used now, we can just remove the orphan in
favor of a well-placed `nub`.

mastapegs commented:
I like the algorithm for this. I think using the past 5 iterations is a good heuristic to gauge point completion per day.
